zookeeper/value_node_watcher.go

```diff
@@ -0,0 +1,318 @@
+package zookeeper
+
+import (
+	"bytes"
+	"sync"
+	"time"
+
+	"github.com/jpillora/backoff"
+	"github.com/samuel/go-zookeeper/zk"
+	"github.com/sirupsen/logrus"
+)
+
+// ValueNodeWatchListener function signature for value node watcher listner, this is used to invoke a callback when a ZK node changes
+type ValueNodeWatchListener func([]byte)
+type valueNodeWatcher struct {
+	client       ZKClient
+	nodePath     string
+	pollTimeout  time.Duration
+	lastVersion  int32
+	value        []byte
+	eventChannel <-chan zk.Event
+	disconnected chan struct{}
+	closed       chan struct{}
+	listener     ValueNodeWatchListener
+	watchActive  bool
+	log          *logrus.Entry
+	watchMutex   sync.Mutex
+}
+
+// CreateValueNodeWatcher returns a ZK ValueNodeWatcher that will call the provided listener when the node value changes, is created or deleted.
+func CreateValueNodeWatcher(client ZKClient, path string, polltimeout time.Duration, listener ValueNodeWatchListener) (ValueNodeWatcher, error) {
+	if listener == nil {
+		return nil, ErrListenerNotProvided
+	}
+	// Ensure client is currently connected, otherwise error
+	if client.ClientState() == Disconnected {
+		return nil, ErrDisconnected
+	}
+	exists, ver, err := client.Exists(path)
+	if err != nil {
+		return nil, ErrFailedToReadNode
+	}
+	var value []byte
+	if exists {
+		value, ver, err = client.Get(path)
+		if err != nil {
+			return nil, ErrFailedToReadNode
+		}
+	} else {
+		value = []byte{}
+	}
+
+	nw := &valueNodeWatcher{
+		client:       client,
+		nodePath:     path,
+		pollTimeout:  polltimeout,
+		lastVersion:  ver,
+		value:        value,
+		eventChannel: nil,
+		disconnected: nil,
+		closed:       make(chan struct{}),
+		listener:     listener,
+		watchActive:  false,
+		log: logrus.WithFields(logrus.Fields{
+			"package": "zookeeper.value_node_watcher",
+			"zk-node": path,
+		}),
+	}
+	nw.log.Debug("Value watcher created.")
```

Comment:
These are OK to keep because you would have to set log level to debug to see them, right? What's the diff between Debug and Trace?

Reply:
Correct, trace is lower than even debug and I tend to use it for very verbose logging, usually when running locally & developing. Debug is data that maybe useful with tracking down an issue. For me honestly the difference is more about feel. and open to change. I just try to start somewhere that makes sense, and then adjust up or down in level as required.